dashboard/app/asset_storage.go

```diff
@@ -0,0 +1,347 @@
+// Copyright 2022 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package main
+
+import (
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/google/syzkaller/dashboard/dashapi"
+	"github.com/google/syzkaller/pkg/asset"
+	"golang.org/x/net/context"
+	"google.golang.org/appengine/v2"
+	db "google.golang.org/appengine/v2/datastore"
+	"google.golang.org/appengine/v2/log"
+)
+
+// TODO: decide if we really want to save job-related assets right now.
+//
+// The problem is that:
+// 1. We cannot first report the job status and then upload the assets -- an email will already
+//    be sent by that time.
+// 2. If we first upload assets and then construct a JobDoneReq response already with assets,
+//    we need to handle extra corner cases:
+//    - If JobDoneReq failed, assets must be deleted.
+//    - If we uploaded several files, all must be deleted.
+//    - What if syz-ci was killed somewhere in between? How do we get rid of those abadoned files?
+//
+// With the existing asset types it does not seem to be worth it to solve those problems.
+// Probably once we have things like code dumps it will start to make some sense.
+
+func saveBuildAsset(c context.Context, ns string, req *dashapi.AddBuildAssetReq) error {
+	asset := Asset{
+		Type:        asset.Type(req.AssetType),
+		DownloadURL: req.DownloadURL,
+		CreateDate:  timeNow(c),
+	}
+	tx := func(c context.Context) error {
+		build, err := loadBuild(c, ns, req.BuildID)
+		if err != nil {
+			return err
+		}
+		build.Assets = append(build.Assets, asset)
+		if _, err := db.Put(c, buildKey(c, ns, req.BuildID), build); err != nil {
+			return fmt.Errorf("failed to put build: %w", err)
+		}
+		log.Infof(c, "updated build: %#v", build)
+		return nil
+	}
+	if err := db.RunInTransaction(c, tx, &db.TransactionOptions{}); err != nil {
+		log.Errorf(c, "failed to update build: %v", err)
```

Comment:
Shouldn't any request errors be already logged somewhere up the stack?
We don't generally log whenever we return an error.
Then this can be just `return db.RunInTransaction(...)`